contracts/twap/CumulativeTwap.sol

```diff
@@ -11,148 +10,208 @@ contract CumulativeTwap is BlockContext {
         }
 
         uint256 currentTimestamp = _blockTimestamp();
         uint256 targetTimestamp = currentTimestamp.sub(interval);
-        (Observation memory beforeOrAt, Observation memory atOrAfter) = _getSurroundingObservations(targetTimestamp);
         uint256 currentCumulativePrice =
             latestObservation.priceCumulative.add(
                 (latestObservation.price.mul(latestUpdatedTimestamp.sub(latestObservation.timestamp))).add(
-                    latestPrice.mul(currentTimestamp.sub(latestUpdatedTimestamp))
+                    price.mul(currentTimestamp.sub(latestUpdatedTimestamp))
                 )
             );
 
-        //
-        //                   beforeOrAt                    atOrAfter
-        //      ------------------+-------------+---------------+------------------
-        //                <-------|             |               |
-        // case 1       targetTimestamp         |               |------->
-        // case 2                               |              targetTimestamp
-        // case 3                          targetTimestamp
-        //
+        // case 1
+        //                                 beforeOrAt     (it doesn't matter)
+        //                              targetTimestamp   atOrAfter
+        //      ------------------+-------------+---------------+----------------->
+
+        // case 2
+        //          (it doesn't matter)     atOrAfter
+        //                   beforeOrAt   targetTimestamp
+        //      ------------------+-------------+--------------------------------->
+
+        // case 3
+        //                   beforeOrAt   targetTimestamp   atOrAfter
+        //      ------------------+-------------+---------------+----------------->
+
+        //                                  atOrAfter
+        //                   beforeOrAt   targetTimestamp
+        //      ------------------+-------------+---------------+----------------->
+
+        (Observation memory beforeOrAt, Observation memory atOrAfter) = _getSurroundingObservations(targetTimestamp);
         uint256 targetCumulativePrice;
-        // case1. not enough historical data or just enough (`==` case)
-        if (targetTimestamp <= beforeOrAt.timestamp) {
-            targetTimestamp = beforeOrAt.timestamp;
+
+        // case1. left boundary
+        if (targetTimestamp == beforeOrAt.timestamp) {
             targetCumulativePrice = beforeOrAt.priceCumulative;
         }
-        // case2. the latest data is older than or equal the request
-        else if (atOrAfter.timestamp <= targetTimestamp) {
-            targetTimestamp = atOrAfter.timestamp;
+        // case2. right boundary
+        else if (atOrAfter.timestamp == targetTimestamp) {
             targetCumulativePrice = atOrAfter.priceCumulative;
         }
+        // not enough historical data
+        else if (beforeOrAt.timestamp == atOrAfter.timestamp) {
+            return 0;
```

Comment:
Same as above, can we just return `price` here?

Reply:
same as above https://github.com/perpetual-protocol/perp-oracle-contract/pull/75/files#r1114107456.

Reply:
updated at f09bf06